chain/pool/txwithtime.go

```diff
@@ -0,0 +1,220 @@
+package pool
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/nknorg/nkn/v2/common"
+	"github.com/nknorg/nkn/v2/pb"
+	"github.com/nknorg/nkn/v2/transaction"
+)
+
+// sync txn
+const MaxSyncTxnInterval = 16000 // in milli seconds
+
+type txnWithTime struct { // abbreviation is twt: txn with time
+	arriveTime int64
+	txn        *transaction.Transaction
+}
+
+// xor of byte slices
+func xorBytes(b1 []byte, b2 []byte) ([]byte, error) {
+	if len(b1) != len(b2) {
+		return nil, fmt.Errorf("Input slices don't have same length")
+	}
+
+	b := make([]byte, len(b1))
+	for i := 0; i < len(b1); i++ {
+		b[i] = b1[i] ^ b2[i]
+	}
+
+	return b, nil
+}
+
+// update twtFingerprint
+func (tp *TxnPool) updateTwtFingerprint(txnHash common.Uint256) []byte {
+
+	tp.fingerprintMu.Lock()
+	defer tp.fingerprintMu.Unlock()
+
+	if tp.twtFingerprint == nil {
+		tp.twtFingerprint = make([]byte, 0, len(txnHash.ToArray()))
+		tp.twtFingerprint = append(tp.twtFingerprint, txnHash.ToArray()...)
+	} else {
+		tp.twtFingerprint, _ = xorBytes(tp.twtFingerprint, txnHash.ToArray())
+	}
+	return tp.twtFingerprint
+}
+
+// when append new txn to pool, we add this txn pointer to txtWithTime map too.
+func (tp *TxnPool) appendTwt(txn *transaction.Transaction) *txnWithTime {
+
+	now := time.Now().UnixMilli()
+
+	twt := &txnWithTime{
+		arriveTime: now,
+		txn:        txn,
+	}
+
+	tp.twtMapMu.Lock()
+	defer tp.twtMapMu.Unlock()
+
+	tp.twtMap.Set(txn.Hash(), twt)
+	tp.updateTwtFingerprint(txn.Hash())
+	tp.twtCount++
+
+	return twt
+}
+
+// remove a txn from twtMap
+func (tp *TxnPool) removeTwt(txnHash common.Uint256) (twt *txnWithTime) {
+	tp.twtMapMu.Lock()
+	defer tp.twtMapMu.Unlock()
+
+	v, ok := tp.twtMap.Delete(txnHash)
+	if ok {
+		tp.twtCount--
+		twt = v.(*txnWithTime)
+		txnHash := twt.txn.Hash()
+		tp.twtFingerprint, _ = xorBytes(tp.twtFingerprint, txnHash.ToArray())
+	}
+
+	return
+}
+
+// romve time over items from txn with time list
+func (tp *TxnPool) RemoveExpiredTwt() (expiredTwt []*txnWithTime) {
```

Comment:
IMO we should also call this function in `requestTxnPoolHashHandler` so sender and receiver call this function at close time

Reply:
yes. added.